lib/handleRequest.js

```diff
@@ -57,19 +57,59 @@ function handleRequest (req, res, params, context) {
   return
 }
 
-function jsonBody (request, reply) {
+function jsonBody (request, reply, options) {
+  const limit = options.limit
+  const contentLength = request.headers['content-length'] === undefined
+    ? NaN
+    : Number.parseInt(request.headers['content-length'], 10)
+
+  if (contentLength > limit) {
+    reply.code(413).send(new Error('Request body is too large'))
+    return
+  }
+
+  var receivedLength = 0
   var body = ''
   var req = request.req
-  req.on('error', onError)
+
   req.on('data', onData)
   req.on('end', onEnd)
-  function onError (err) {
-    reply.code(422).send(err)
-  }
+  req.on('error', onEnd)
+
   function onData (chunk) {
-    body += chunk
+    receivedLength += chunk.length
+
+    if (receivedLength > limit) {
+      req.removeListener('data', onData)
+      req.removeListener('end', onEnd)
+      req.removeListener('error', onEnd)
+      reply.code(413).send(new Error('Request body is too large'))
+      return
+    }
+
+    body += chunk.toString()
```

Comment:
`chunk + ''` performs better

Reply:
In my benchmarks that performs exactly the same as `body += chunk`
(I tried on both Node 8.9.4 and 9.3.0)